File: apps/shadow-lens/unity/Assets/ShadowLens/VoiceV2/ShadowVoiceQueue.cs
// apps/shadow-lens/unity/Assets/ShadowLens/VoiceV2/ShadowVoiceQueue.cs
// Priority playback queue + barge-in (mirror of lib/voice/shadow-voice-queue.mjs). Higher priority
// interrupts lower; an ordinary status never interrupts an in-progress verbatim quote; duplicates are
// suppressed; reset/language-switch clear stale utterances; queue length is bounded. Pure C#.
using System.Collections.Generic;

namespace ShadowLens.VoiceV2
{
    public sealed class ShadowVoiceQueue
    {
        readonly List<SpokenUtterance> _queue = new List<SpokenUtterance>();
        public SpokenUtterance Current { get; private set; }
        public int Dropped { get; private set; }
        public int Suppressed { get; private set; }
        // set true by the player while a verbatim-quote segment of Current is speaking
        public bool CurrentActiveIsVerbatimQuote;
        const int MaxQueue = 64;

        public struct EnqueueResult { public bool Accepted; public SpokenUtterance Interrupted; }

        public EnqueueResult Enqueue(SpokenUtterance u)
        {
            if ((Current != null && Current.UtteranceId == u.UtteranceId) || _queue.Exists(q => q.UtteranceId == u.UtteranceId))
            { Suppressed++; return new EnqueueResult { Accepted = false, Interrupted = null }; }
            SpokenUtterance interrupted = null;
            if (Current != null && ShouldInterrupt(u, Current)) { interrupted = Current; Current = null; }
            _queue.Add(u);
            _queue.Sort((a, b) => Rank(a.Priority).CompareTo(Rank(b.Priority)));
            while (_queue.Count > MaxQueue) { _queue.RemoveAt(_queue.Count - 1); Dropped++; }
            return new EnqueueResult { Accepted = true, Interrupted = interrupted };
        }

        bool ShouldInterrupt(SpokenUtterance incoming, SpokenUtterance current)
        {
            if (current.Interruptibility == Interruptibility.NON_INTERRUPTIBLE) return false;
            if (CurrentActiveIsVerbatimQuote && Rank(incoming.Priority) > Rank(VoicePriority.P0)) return false; // only P0 interrupts a quote
            return Rank(incoming.Priority) < Rank(current.Priority);
        }

        public SpokenUtterance Next() { if (Current == null && _queue.Count > 0) { Current = _queue[0]; _queue.RemoveAt(0); } return Current; }
        public SpokenUtterance UserInterrupt() { var c = Current; Current = null; CurrentActiveIsVerbatimQuote = false; return c; }
        public bool StopAll() { bool had = Current != null || _queue.Count > 0; Current = null; _queue.Clear(); CurrentActiveIsVerbatimQuote = false; return had; }
        public void ClearLocaleExcept(string locale) { _queue.RemoveAll(u => u.Locale != locale); if (Current != null && Current.Locale != locale) Current = null; }
        public void OnPause() { Current = null; }
        public int Length => _queue.Count;

        static int Rank(VoicePriority p) => (int)p;
    }
}

File: lib/voice/shadow-voice-queue.mjs
// Priority playback queue + barge-in rules. Pure + deterministic (no timers here — the host drives
// tick()). Priorities P0(highest)..P4. A higher-priority utterance interrupts a lower one; an ordinary
// status update never interrupts an in-progress verbatim evidence quote; duplicates are suppressed;
// a language switch or reset clears stale utterances; the queue length is bounded.
const RANK = { P0: 0, P1: 1, P2: 2, P3: 3, P4: 4 };
const MAX_QUEUE = 64;

export class ShadowVoiceQueue {
  constructor() { this.queue = []; this.current = null; this.dropped = 0; this.suppressed = 0; }

  // Try to enqueue. Returns { accepted, interrupted } — interrupted is the utterance that was barged.
  enqueue(u) {
    // duplicate suppression: same utterance_id already current or queued
    if ((this.current && this.current.utterance_id === u.utterance_id) || this.queue.some((q) => q.utterance_id === u.utterance_id)) {
      this.suppressed++; return { accepted: false, interrupted: null };
    }
    let interrupted = null;
    if (this.current && this.shouldInterrupt(u, this.current)) { interrupted = this.current; this.current = null; }
    this.queue.push(u);
    this.queue.sort((a, b) => RANK[a.priority] - RANK[b.priority]);
    while (this.queue.length > MAX_QUEUE) { this.queue.pop(); this.dropped++; }
    return { accepted: true, interrupted };
  }

  // Barge-in policy: a strictly higher priority interrupts; but nothing ordinary interrupts a
  // NON_INTERRUPTIBLE current OR a current whose active segment is a verbatim quote.
  shouldInterrupt(incoming, current) {
    if (current.interruptibility === "NON_INTERRUPTIBLE") return false;
    if (current._activeIsVerbatimQuote && RANK[incoming.priority] > RANK.P0) return false; // only P0 safety interrupts a quote
    return RANK[incoming.priority] < RANK[current.priority];
  }

  // Advance: if nothing is playing, take the highest-priority queued item.
  next() { if (!this.current && this.queue.length) this.current = this.queue.shift(); return this.current; }

  // User barge-in (an explicit interaction) stops the current utterance.
  userInterrupt() { const c = this.current; this.current = null; return c; }

  // Back / Cancel / Reset stop speech and clear the queue.
  stopAll() { const had = this.current || this.queue.length; this.current = null; this.queue = []; return !!had; }

  // Language switch cancels utterances of the old locale (keep only the new one).
  clearLocaleExcept(locale) {
    this.queue = this.queue.filter((u) => u.locale === locale);
    if (this.current && this.current.locale !== locale) this.current = null;
  }

  // App pause: stop current; keep nothing obsolete to replay.
  onPause() { this.current = null; }

  get length() { return this.queue.length; }
}
